Added Firebase analytics

max2me opened the pull request.

max2me committed:
Added Firebase dependencies.

max2me committed:
Selectively including GMS dependencies to reduce app size, adding multidex support, fixing deprecated APIs

max2me committed:
Adding toggling collecting data on and off

max2me committed:
Added instrumentation for various activities

max2me committed:
Added dialog to prompt Play Services update

lognaturel commented:
Looks great! 🎉🎉 

Few questions/comments:
- Does WAKE_LOCK permission need to be explicitly added to the manifest? [This](http://stackoverflow.com/questions/33871981/does-google-analytics-add-wakelock-permission) suggests analytics automatically sets it and I think it'd be a bit nicer not to have the explicit permission if possible since the app itself doesn't require it. Or I suppose having it explicitly with a comment making it clear it's for analytics could be good too.
- I think the `analytics` package would be better named as `org.odk.collect.analytics`
- The `instance` singleton leaks the context since `Analytics` holds a reference to it. Instead of passing in the context and holding a reference, you should be able to use `getApplicationContext()` as you need it.
- Is `EventTypes.java` for an enum that will represent events at some point? Even if it's blank for now, a quick comment might be nice to have and maybe while you're there just do a single example so it's clear to someone else who might want to log events.
- I didn't test your small updates to the mapping activities but they look like good equivalents. That ended up being required because of API updates, right?
- For consistency, I think 'allow us to prioritise' should be the American way with 'prioritize' (I'm not saying one is more 'correct' than the other 😉)
- I noticed a number of tab/space inconsistencies. I'll mark a couple. It looks like the project doesn't follow a consistent convention so maybe just be sure to match surrounding code for now. This should probably be standardized at some point.
- If the `ensureGooglePlayServicesIsUpToDate` test fails, there'll just be some kind of error message saying Google Play Services isn't installed/updated and then the app will work as normal, right? Is this actionable to the user? I guess maybe they could update or install Play Services if they need to? Also, since I believe the `ConnectionResult` test is exhaustive, I'd probably write it as `if (!ConnectionResult.SUCCESS) show the error` instead of using a switch.
- Small thing and you can decide whether you like this or not but I'd tend to put the `getScreenID` method in the `ScreenType enum` so related functionality is grouped. You can use `switch (this)`.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -1255,6 +1063,216 @@ public void onClick(View v) {
+		InputFilter returnFilter = new InputFilter() {
+            public CharSequence filter(CharSequence source, int start,
+									   int end, Spanned dest, int dstart, int dend) {
+                for (int i = start; i < end; i++) {
```
Some strange things happened to indentation and line breaks throughout these new methods!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FileManagerTabs.java:
```diff
@@ -82,6 +92,18 @@ protected void onCreate(Bundle savedInstanceState) {
 	}
 
+	private void logScreenView(String tabId) {
+		switch (tabId) {
```
Indentation isn't consistent.

max2me committed:
Merge remote-tracking branch 'opendatakit/master' into feature-analytics

max2me committed:
Added comment for WAKE_LOCK permissions describing dev team stance on it

max2me committed:
Moved nalytics package into org.odk.collect.android

max2me committed:
Start using getApplicationContext in Analytics

max2me committed:
Updated copy of the message about stats collected

max2me committed:
Removed empty EventTypes

max2me commented:
@lognaturel Hélène, thanks again for great feedback!

**Wake_lock permission**
It is indeed part of the implementation and Firebase team suggests keeping it. I've added a comment and a link to the thread.

**Renaming `analytics` package**
Done!

**Leaking context**
I'm not 100% sure what you mean there so for now I changed it to using getApplicationContext() and at least [this thread indicates](http://stackoverflow.com/questions/36817412/singleton-with-context-as-a-variable-memory-leaks) that it shouldnt leak anything. I'm happy to modify it to pass context every time we try to initialize it or log a screen view.

**EventTypes.java**
I should pay more attention and not leave files like that. Removed it, thank you for pointing it out.

**Map activities updates**
Correct. As we've updated to newer version of Play Services old synchronous APIs for Map fragments got removed.

**Updating copy**
Done!

**ensureGooglePlayServicesIsUpToDate()**
So if Play services are outdated, disabled or not installed, appropriate dialog will be shown encouraging user to update, enable or install them ([per google documentation](https://developers.google.com/android/guides/setup#ensure_devices_have_the_google_play_services_apk)). Until user performs that step she wont be able to use the app (I tested it by reverting Play Services on my phone to old version - I do not recommend it as it seems to mess up push notifications for all my apps).  

We could remove dialog or show our own version of it to avoid preventing users from using the app but I'm a little concerned that it can lead to unexpected crashes caused by Firebase dependencies (dont have evidence for this, just a gut feeling).  

`ConnectionResult` contains more results than what's listed in `switch` block and so I feel like checking just for Success one will be a bit less clear in this case although if you feel strongly, I'm happy to make that change.

**getScreenID() method**
I can see rationale for both. On one hand it's nice to that string ID closer to the actual enum, on another it's only used when recording event and no other classes use that. Lmk if you feel strong about this change and I'll incorporate it (left as-is for now).

max2me commented:
**Formatting inconsistencies**  
I havent addressed inconsistency issues yet, mostly because I want to do it on the project level. As you mentioned, right now tabs/space are a bit all over the place. Once all active pull requests are merged, I was hoping to use automatic formatting in Android Studio on all Java files and fix it everywhere.

Let me know what you think about it and what should be default setting for this project (tabs vs space etc.) 

Thanks again for the feedback!

lognaturel commented on collect_app/src/main/AndroidManifest.xml:
```diff
@@ -31,6 +31,11 @@
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
+    <!--
```
I think this comment is the right idea but I think you may have accidentally added the wrong one! For wake lock, the explanation is here: https://groups.google.com/d/msg/firebase-talk/CXgecSxgsRE/avud-Rt7AwAJ -- basically, "without WAKE_LOCK, analytics data loss will occur."

Reading that thread and what you added in the comment suggests that the RECEIVE and C2D_MESSAGE permissions are also required, though. Is that what you're understanding as well?

max2me replied:
You are right! I've thought that his comment applied to all permissions but after reading thread it's obvious that it might not be the case. I'll correct it.

lognaturel commented:
I believe your fix for the `context` leak works. You just can't save a direct reference to context and instead need to call `getApplicationContext()` but I guess it's ok to save that value. I'd tend to call `getApplicationContext()` in `Analytics` like the StackOverflow snippet but I suppose it's equivalent!

I added a brief note about permissions inline.

Doing the indentation fixes in one pass sounds fine. I would suggest following the [Android contributers' guidelines](https://source.android.com/source/code-style.html#use-spaces-for-indentation) -- 4 spaces for regular indentation and 8 spaces for hanging indents.

The play services situation is going to require a community conversation I think. Making new versions of the app unusable for devices without play services installed is probably not an option. Even if that's overcome, the question becomes whether analytics are meaningful if phones without play services can't be counted. Anyone with opinions should certainly weigh in here and @yanokwa can also think about how to facilitate a broader conversation.

In the mean time, I'd say fix the permissions thing so everything else will be ready pending a community decision. 👍

max2me committed:
Corrected comment about wake_lock permissions required for analytics

max2me committed:
Merge remote-tracking branch 'opendatakit/master' into feature-analytics

lognaturel commented:
Closing this PR for now as the conversation around analytics continues.
